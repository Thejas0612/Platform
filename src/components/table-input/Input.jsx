import { TextField } from "@mui/material";
import { useState } from "react";
import CustomTooltip from "./CustomTooltip";
import { TEN_THOUSAND_PLACE_DECIMAL_MATCHER } from "../../utils/constants";

const InputProps = {
  style: {
    minWidth: "5rem"
  }
};

const Input = (props) => {
  const [defaultValue, setDefaultvalue] = useState("");
  const [error, setError] = useState(false);
  const [errorMessage, setErrorMessage] = useState("");

  const handleChange = (e) => {
    const value = e.target.value;
    
      setDefaultvalue(value);
      if (props.schemaProps.min != null && value !== "" && value < props.schemaProps.min) {
        setError(true);
        setErrorMessage(props.schemaProps.minError);
        return;
      } else {
        setError(false);
        setErrorMessage("");
      }
      props.onChange(e, e.target.value, props.schemaProps.name);
  };
  const precise = (x) => {
    if(props?.schemaProps?.precision){
      return isNaN(Number.parseFloat(x).toFixed(props?.schemaProps?.precision)) ? "" : Number.parseFloat(x).toFixed(props?.schemaProps?.precision) ;
    }
      return x;
  }
  return (
    <CustomTooltip title={error ? errorMessage : ""} placement="top">
      <TextField
        value={defaultValue}
        onChange={handleChange}
        onBlur={(e) => setDefaultvalue( defaultValue !== "" && defaultValue !== "-" && String(precise(defaultValue)) !== "" ?   String(precise(defaultValue)).split(".")[0].substring(0,6)+"."+String(precise(defaultValue)).split(".")[1].substring(0,4)  : ""  )}
        size={props.size}
        error={error}
        InputProps={{
          ...InputProps,
          className: props.schemaProps.required ? "Mui-focused " : ""
        }}
        disabled={props.schemaProps.disabled}
      />
    </CustomTooltip>
  );
};

export default Input;
