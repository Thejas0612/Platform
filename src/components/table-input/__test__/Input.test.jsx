import { fireEvent, render, screen } from "@testing-library/react";
import Input from "../Input";
import userEvent from "@testing-library/user-event";

describe("<Input />", () => {
  const schemaProps = {
    type: "TEXT_INPUT",
    name: "density_norm",
    inputClass: "customRequired",
    disabled: false,
    required: true,
    align: "center",
    precision: 4,
    min: 0.0001,
    minError: "Entered Normal Density is below or equal to 0"
  };

  test("Show error when the value is less than the minimum", async () => {
    const { getByRole, container } = render(<Input schemaProps={schemaProps} size="small" />);

    const input = getByRole("textbox");
    fireEvent.change(input, { target: { value: "-1" } });
    await userEvent.hover(input);
    const tooltip = await screen.findByRole("tooltip");

    expect(tooltip).toMatchSnapshot();
    expect(container).toMatchSnapshot();
  });

  test("Does not show error if the value is above the minimum", async () => {
    const { getByRole, container } = render(<Input schemaProps={schemaProps} size="small" />);

    const input = getByRole("textbox");
    fireEvent.change(input, { target: { value: "5" } });
    fireEvent.blur(input);

    expect(container).toMatchSnapshot();
  });

  test("it should not contain a value in value if a - is typed", async () => {
    const { getByRole, container } = render(<Input schemaProps={schemaProps} size="small" />);

    const input = getByRole("textbox");
    fireEvent.change(input, { target: { value: "-" } });
    fireEvent.blur(input);

    expect(container).toMatchSnapshot();
  });
});
