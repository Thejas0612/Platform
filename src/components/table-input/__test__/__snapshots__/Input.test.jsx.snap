// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> Does not show error if the value is above the minimum 1`] = `
<div>
  <div
    aria-label=""
    class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
    data-mui-internal-clone-element="true"
  >
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall Mui-focused  css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
      style="min-width: 5rem;"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
        id="mui-3"
        type="text"
        value="5.0000"
      />
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-ihdtdm"
        >
          <span
            class="notranslate"
          >
            ​
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;

exports[`<Input /> Show error when the value is less than the minimum 1`] = `
<div
  class="base-Popper-root MuiTooltip-popper MuiTooltip-popperInteractive css-8v5zyd-MuiPopper-root-MuiTooltip-popper"
  data-popper-escaped=""
  data-popper-placement="top"
  data-popper-reference-hidden=""
  id="mui-2"
  role="tooltip"
  style="position: absolute; top: 0px; left: 0px; margin: 0px; bottom: 0px; transform: translate(0px, 0px);"
>
  <div
    class="MuiTooltip-tooltip MuiTooltip-tooltipPlacementTop css-ja5taz-MuiTooltip-tooltip"
    style="opacity: 1; transform: scale(1, 1); transition: opacity 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 133ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
  >
    Entered Normal Density is below or equal to 0
  </div>
</div>
`;

exports[`<Input /> Show error when the value is less than the minimum 2`] = `
<div>
  <div
    aria-label="Entered Normal Density is below or equal to 0"
    class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
    data-mui-internal-clone-element="true"
  >
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-error MuiInputBase-formControl MuiInputBase-sizeSmall Mui-focused  css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
      style="min-width: 5rem;"
    >
      <input
        aria-invalid="true"
        class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
        id="mui-1"
        type="text"
        value="-1"
      />
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-ihdtdm"
        >
          <span
            class="notranslate"
          >
            ​
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;

exports[`<Input /> it should not contain a value in value if a - is typed 1`] = `
<div>
  <div
    aria-label=""
    class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
    data-mui-internal-clone-element="true"
  >
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall Mui-focused  css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
      style="min-width: 5rem;"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
        id="mui-5"
        type="text"
        value=""
      />
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-ihdtdm"
        >
          <span
            class="notranslate"
          >
            ​
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;
